yoyodyne/models/hard_attention.py

```diff
@@ -0,0 +1,461 @@
+"""Hard monotonic neural HMM classes."""
+
+import argparse
+from typing import Callable, Dict, Optional, Tuple
+
+import numpy
+import torch
+from torch import nn
+
+from .. import data, defaults, util
+from . import lstm, modules
+
+NEG_LOG_EPSILON = -numpy.log(1e-7)
```

Comment:
why not just use `math.log` for this? There's no real need for `numpy` is there?

Reply:
Isn't numpy more stable than `math`? Especially since it's being cast for a tensor operation

Reply:
I don't have any objection to using `numpy` in general, but here they produce exactly the same value, a `float` scalar which are exactly the same value: 

```python
assert numpy.log(1e7) == math.log(1e7)
```

I think for scalars they both are just calling into the same `libc` function anyways.

Reply:
I'd rather just use numpy for consistency with other libraries. (I see everyone use the numpy log). It's the same number of imports and we need to keep support with the dependency anyhow.

Reply:
Internal consistency is something I care about, and we've used `math` [elsewhere in the library](https://github.com/CUNY-CL/yoyodyne/blob/c0640676adc8e25b3e41b5f7c574e7e115381a03/yoyodyne/models/pointer_generator.py#L20), so I will either fix that, or this, at some later date. Go ahead and I'll resolve issue. 

(I don't care about "external consistency" at all.)

Reply:
| (I don't care about "external consistency" at all.)
I do, it's easier when adapting codebases and sanity checking stability issues. So i'd recommend just changing occurrences of `math.log` to `numpy.log`

Reply:
okay do you want to do that in this PR or should I later?

Reply:
Sure, lemme search and find all instances.